src/Symfony/Component/Security/Core/Signature/SignatureHasher.php

```diff
@@ -45,24 +45,48 @@ public function __construct(PropertyAccessorInterface $propertyAccessor, array $
     }
 
     /**
-     * Verifies the hash using the provided user and expire time.
+     * Verifies the hash using the provided user identifier and expire time.
+     *
+     * This method must be called before the user object is loaded from a provider.
      *
      * @param int    $expires The expiry time as a unix timestamp
      * @param string $hash    The plaintext hash provided by the request
      *
      * @throws InvalidSignatureException If the signature does not match the provided parameters
      * @throws ExpiredSignatureException If the signature is no longer valid
      */
-    public function verifySignatureHash(UserInterface $user, int $expires, string $hash): void
+    public function acceptSignatureHash(string $userIdentifier, int $expires, string $hash): void
     {
-        if (!hash_equals($hash, $this->computeSignatureHash($user, $expires))) {
+        if ($expires < time()) {
+            throw new ExpiredSignatureException('Signature has expired.');
+        }
+        $hmac = substr($hash, 0, 44);
+        $payload = substr($hash, 44).':'.$expires.':'.$userIdentifier;
+
+        if (!hash_equals($hmac, strtr(base64_encode(hash_hmac('sha256', $payload, $this->secret, true)), '+/=', '-_~'))) {
```

Comment:
it might make sense to implement this hash computation in a private method instead of duplicating it between this verification method and `computeSignatureHash`.